File: MergeSort/mergesort.java
import java.util.*;

class Merge_Sort
{
    ArrayList<Integer> arr=new ArrayList<>();
    public Merge_Sort(ArrayList<Integer> array)
    {
        this.arr=array;
    }

    public void merging(int start,int mid,int end)
    {
       ArrayList<Integer> left = new ArrayList<>(arr.subList(start, mid+1));         // slice the arr from start index to mid index
       ArrayList<Integer> right = new ArrayList<>(arr.subList(mid+1, end+1));        // slice the arr from mid+1 index to end index
       
       int x=0,y=0;

       for(int i=start;i<=end;i++)
       {
           if(left.get(x)<right.get(y))
           {
               arr.set(i,left.get(x));
               x++;
           }
           else 
           {
               arr.set(i,right.get(y));
               y++;
           }

           if(x==left.size())
           {
              while(y<right.size())
              {
                 i++;
                 arr.set(i,right.get(y));
                 y++;
              }
              break;
           }
           else if(y == right.size())
           {
               while(x<left.size())
               {
                i++;
                arr.set(i,left.get(x));
                x++;
               }
               break;
               
           }

       }
    }

    public void mergesort(int start,int end)
    {
        if (start == end)
        {
            return;
        }
        int mid=(int) (start+end)/2;
        mergesort(start,mid);
        mergesort(mid+1,end);
        merging(start,mid,end);
    }

    

    public void printArr()
    {
        System.out.println(arr);
    }
    
}

class mergesort
{  
      public static void main(String[] args)
    {
        ArrayList<Integer> arr=new ArrayList<>();
        char ch='Y';
        int ele;
        Scanner sc=new Scanner(System.in);
        while(ch!='N' && ch!='n')
        {
            System.out.print("Enter the element:");
            ele=sc.nextInt();
            arr.add(ele);
            System.out.print("Continue ? Y or N: ");
            ch=sc.next().charAt(0);
        }
       
        Merge_Sort msort=new Merge_Sort(arr);

        System.out.println("ArrayList before merge sort: ");
        msort.printArr();

        msort.mergesort(0,arr.size()-1);

        System.out.println("ArrayList after merge sort: ");
        msort.printArr();

    }
}


/*
OUTPUT:
Enter the element:65
Continue ? Y or N: y
Enter the element:34
Continue ? Y or N: y
Enter the element:90
Continue ? Y or N: y
Enter the element:1
Continue ? Y or N: y
Enter the element:23
Continue ? Y or N: y
Enter the element:96
Continue ? Y or N: y
Enter the element:5
Continue ? Y or N: y
Enter the element:76
Continue ? Y or N: n
ArrayList before merge sort: 
[65, 34, 90, 1, 23, 96, 5, 76]
ArrayList after merge sort: 
[1, 5, 23, 34, 65, 76, 90, 96]

*/

File: MergeSort/mergesort.py

class MergeSort:
    
    Arr=[]
    def __init__(self,arr):
        self.Arr=arr
        
    def merge(self,start,mid,end):
        Arr=self.Arr
        left=Arr[start:mid+1]
        right=Arr[mid+1:end+1]
        x,y=0,0
        for i in range(start,end+1):
            if left[x]<right[y]:
                Arr[i]=left[x]
                x+=1
            else:
                Arr[i]=right[y]
                y+=1
            if x==len(left):
                while y<len(right):
                    i+=1
                    Arr[i]=right[y]
                    y+=1
                break
            elif y==len(right):
                while x<len(left):
                    i+=1
                    Arr[i]=left[x]
                    x+=1
                break
        self.Arr=Arr
                
    def merge_sort(self,start,end):
        if start==end:
            return
        mid=(start+end)//2
        self.merge_sort(start,mid)
        self.merge_sort(mid+1,end)
        self.merge(start,mid,end)
    
    def print(self):
        print("List after merge sorting: ",end='')
        for i in self.Arr:print(i,end=" ")
        print()

class Main:
    def main():
        print("Enter the list to sort: ",end='')
        li=[int(i) for i in input().split(" ")]
        msort=MergeSort(li)
        msort.merge_sort(0,len(li)-1)
        msort.print()
        
    if __name__=="__main__":
        main()
        

        
"""
OUTPUT:

Enter the list to sort: 69 78 45 23 89 1 56 99 34 7
List after merge sorting: 1 7 23 34 45 56 69 78 89 99 

"""
